lepton/qemu.go

```diff
@@ -7,37 +7,112 @@ import (
 	"strings"
 )
 
+const qemuBaseCommand = "qemu-system-x86_64"
+
+type QemuRender interface {
+	render() []string
+}
+
 type drive struct {
 	path   string
 	format string
+	iftype string
+	index  string
+}
+
+type device struct {
+	driver   string
+	mac      string
+	netdevid string
 }
 
 type netdev struct {
-	nettype string
-	id      string
-	ifname  string
-	mac     string
+	nettype    string
+	id         string
+	ifname     string
+	script     string
+	downscript string
+	hports     []portfwd
+}
+
+type portfwd struct {
+	port  int
+	proto string
+}
+
+type display struct {
+	disptype string
+}
+
+type serial struct {
+	serialtype string
 }
 
 type qemu struct {
-	cmd    *exec.Cmd
-	drives []drive
-	ifaces []netdev
+	cmd     *exec.Cmd
+	drives  []drive
+	devices []device
+	ifaces  []netdev
+	display display
+	serial  serial
 }
 
-func (q *qemu) addDrive(path string, format string) {
+func (d *display) render() []string {
+	return []string{"-display", fmt.Sprintf("%s", d.disptype)}
+}
 
+func (s *serial) render() []string {
+	return []string{"-serial", fmt.Sprintf("%s", s.serialtype)}
 }
 
-func (q *qemu) addNetworkDevice(n netdev) {
+func (d *drive) render() []string {
+	str := fmt.Sprintf("file=%s,format=%s", d.path, d.format)
+	if len(d.index) > 0 {
+		str = str + "," + fmt.Sprintf("index=%s", d.index)
+	}
+	if len(d.iftype) > 0 {
+		str = str + "," + fmt.Sprintf("if=%s", d.iftype)
+	}
+	return []string{"-drive", str}
+}
 
+func (dv *device) render() []string {
+	str := fmt.Sprintf("%s,netdev=%s", dv.driver, dv.netdevid)
+	if len(dv.mac) > 0 {
+		str = str + "," + fmt.Sprintf("mac=%s", dv.mac)
+	}
+	return []string{"-device", str}
+}
+
+func (nd *netdev) render() []string {
+	var str string
+	str = fmt.Sprintf("%s,id=%s", nd.nettype, nd.id)
+	if len(nd.ifname) > 0 {
+		str += "," + fmt.Sprintf("ifname=%s", nd.ifname)
+	}
+	if len(nd.script) > 0 {
+		str += "," + fmt.Sprintf("script=%s", nd.script)
+	}
+	if len(nd.downscript) > 0 {
```

Comment:
If there is no downscript you need to explicitly set it to 'no' to avoid qemu trying to execute script from a default location and then asking to be root.